### ISSUE TITLE
Is it possible to automatically accept a certificate?

### ISSUE THREAD

**Issue by RoxKilly** [PHPMailer/PHPMailer#540]
Thanks for this great tool. I've just discovered it and am learning. 

I run PHP on a windows machine and want to send encrypted email out through my own SMTP web server. The server supports TLS encryption, but because the certificate is self-signed it gets flagged by pretty much every application that connects to it. For instance, in Outlook, I have to manually accept the certificate to connect; same in Android.

When I use PHPMailer, the SMTP connection fails with the error `Fatalstream_socket_enable_crypto(): SSL operation failed with code 1. OpenSSL Error messages:
error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed' in SMTP.class.php:346` This is in `SMTP::startTLS()`.  

I could disable encryption but I'd like to avoid that so my question is simple: **Is there a setting to automatically accept a self-signed certificate when connecting to SMTP?**

PHP level: beginner
PHPMailer: 1st day

**Comment by ooxi** [comment]
Yes you can disable the certification checking process. But be aware disabling checking the certificate is nearly equivalent to disabling encryption altogether. Imo it's even worse since it gives the user a false feeling of security whereas a plain text connection is obviously insecure.

**Comment by RoxKilly** [comment]
@ooxi,

Thanks for the response and the security consideration; I get your point. Your response brings to mind two questions.

1) How to disable certificate checking? You said it can be done without telling me how :-)

2) Is there a middle ground? In other words, if certificate validation fails, only accept certificates from a supplied whitelist.  This would allow for a secure connection since I know I can trust a particular certificate but would not be exposed to other invalid certs.  This is how some email clients handle the situation. My Android phone and computer will accept the certificate from my server even though it's self-signed but will not accept other self-signed or invalid certs unless I add them to the whitelist. Same thing with the FTP server I use to connect to the server.

**Comment by ooxi** [comment]
[The PHPMailer wiki](https://github.com/PHPMailer/PHPMailer/wiki/Troubleshooting#php-56-certificate-verification-failure) contains information about how to disable certificate checking. Please refer to the [PHP SSL documentation](http://php.net/manual/en/context.ssl.php) for extended information about the meaning of the individual options.

For 1) you are looking for `verify_peer` and `verify_peer_name`, for 2) you are looking for `cafile` :)

**Comment by RoxKilly** [comment]
Thanks I'll take a look!

**Comment by ooxi** [comment]
You're welcome :) please remember to close the ticket as soon as you feel confident your question was answered.

**Comment by RoxKilly** [comment]
You've pointed me in the right direction. I'm able to resolve my point 1 above by adding:
> $mailer->SMTPOptions = ['ssl'=> ['allow_self_signed' => true]];

This worked! Now I wanted to take your advice and strengthen the security by white-listing the certificate I actually expected. So, as instructed by your [2nd link](http://php.net/manual/en/context.ssl.php), I've tried to make use of the `peer_fingerprint` option by modifying my code to the following:
>$mailer->SMTPOptions = ['ssl'=>[
                'allow_self_signed' => true,
                'peer_fingerprint'=>['sha1'=>'751608908fd7361258827e081c136cb4ae73df4c']
            ]];

This however throws the puzzling error `non-Fatal stream_socket_enable_crypto(): Expected peer fingerprint must be a string or an array`.  I tried changing the peer fingerprint value to a string:
>$mailer->SMTPOptions = ['ssl'=>[
                'allow_self_signed' => true,
                'peer_fingerprint'=>'751608908fd7361258827e081c136cb4ae73df4c'
            ]];

The same error is thrown. This is quite strange. `peer_fingerprint` was added with PHP 5.6 and I am using PHP 5.6.13 on Windows, so I should be set.
